--- xorg-server-1.16.3/os/Makefile.am~	2015-01-04 14:05:54.000000000 +0000
+++ xorg-server-1.16.3/os/Makefile.am	2015-01-04 14:23:50.646159253 +0000
@@ -52,12 +52,12 @@
 
 if SPECIAL_DTRACE_OBJECTS
 # Generate dtrace object code for probes in libos & libdix
-dtrace.o: $(top_srcdir)/dix/Xserver.d $(am_libos_la_OBJECTS)
+dtrace.o: $(top_srcdir)/dix/Xserver.d $(am_libos_la_OBJECTS) $(LIBOBJS)
 	$(AM_V_GEN)$(DTRACE) -G -C -o $@ -s $(top_srcdir)/dix/Xserver.d .libs/*.o ../dix/.libs/*.o
 
 noinst_PROGRAMS = os.O
 
 os_O_SOURCES =
-os.O: dtrace.o $(am_libos_la_OBJECTS)
+os.O: dtrace.o $(am_libos_la_OBJECTS) $(LIBOBJS)
 	$(AM_V_GEN)ld -r -o $@ dtrace.o .libs/*.o
 endif
