--- xorg-server-1.11.3/hw/kdrive/linux/keyboard.c.org	2011-12-17 08:32:26.955371326 +0100
+++ xorg-server-1.11.3/hw/kdrive/linux/keyboard.c	2011-12-17 08:32:45.179135298 +0100
@@ -692,7 +692,7 @@
         return !Success;
 
     fd = LinuxConsoleFd;
-    ki->driverPrivate = (void *) fd;
+    ki->driverPrivate = (void *) (intptr_t) fd;
 
     ioctl (fd, KDGKBMODE, &LinuxKbdTrans);
     tcgetattr (fd, &LinuxTermios);
@@ -724,7 +724,7 @@
     if (!ki)
         return;
 
-    fd = (int) ki->driverPrivate;
+    fd = (int) (intptr_t) ki->driverPrivate;
 
     KdUnregisterFd(ki, fd, FALSE);
     ioctl(fd, KDSKBMODE, LinuxKbdTrans);
@@ -753,7 +753,7 @@
     if (!ki)
         return;
 
-    ioctl ((int)ki->driverPrivate, KDSETLED, leds & 7);
+    ioctl ((int) (intptr_t)ki->driverPrivate, KDSETLED, leds & 7);
 }
 
 KdKeyboardDriver LinuxKeyboardDriver = {
--- xorg-server-1.11.3/hw/kdrive/linux/ms.c.org	2011-12-17 08:33:00.929507448 +0100
+++ xorg-server-1.11.3/hw/kdrive/linux/ms.c	2011-12-17 08:33:21.106650997 +0100
@@ -152,7 +152,7 @@
     }
     if (KdRegisterFd (port, MsRead, pi))
 	return TRUE;
-    pi->driverPrivate = (void *)port;
+    pi->driverPrivate = (void *) (intptr_t)port;
 
     return Success;
 
@@ -164,7 +164,7 @@
 static void
 MsDisable (KdPointerInfo *pi)
 {
-    KdUnregisterFd (pi, (int)pi->driverPrivate, TRUE);
+    KdUnregisterFd (pi, (int)(intptr_t)pi->driverPrivate, TRUE);
 }
 
 static void
--- xorg-server-1.11.3/hw/kdrive/linux/ps2.c.org	2011-12-17 08:33:34.930310934 +0100
+++ xorg-server-1.11.3/hw/kdrive/linux/ps2.c	2011-12-17 08:33:53.580751546 +0100
@@ -159,7 +159,7 @@
         return BadAlloc;
     }
 
-    pi->driverPrivate = (void *)fd;
+    pi->driverPrivate = (void *)(intptr_t)fd;
 
     return Success;
 }
@@ -168,7 +168,7 @@
 static void
 Ps2Disable (KdPointerInfo *pi)
 {
-    KdUnregisterFd (pi, (int)pi->driverPrivate, TRUE);
+    KdUnregisterFd (pi, (int)(intptr_t)pi->driverPrivate, TRUE);
 }
 
 static void
