--- configure.ac~	2009-08-01 11:18:31.911298999 +0200
+++ configure.ac	2009-08-01 11:56:36.783570379 +0200
@@ -599,7 +599,7 @@
 dnl first and if it's not in the default location, that you set the ACLOCAL
 dnl environment variable to find it, such as:
 dnl	ACLOCAL="aclocal -I ${PREFIX}/share/aclocal"
-m4_pattern_forbid(XTRANS_CONNECTION_FLAGS)
+dnl m4_pattern_forbid(XTRANS_CONNECTION_FLAGS)
 
 # Transport selection macro from xtrans.m4
 XTRANS_CONNECTION_FLAGS
