--- xorg-server-1.8.2/hw/xfree86/common/xf86cmap.c~	2010-07-01 06:06:35.000000000 +0200
+++ xorg-server-1.8.2/hw/xfree86/common/xf86cmap.c	2010-07-17 11:09:43.640632480 +0200
@@ -144,9 +144,11 @@
     int *indices; 
     int elements;
 
+#if 0
     /* If we support a better colormap system, then pretend we succeeded. */
     if (xf86_crtc_supports_gamma(pScrn))
 	return TRUE;
+#endif
 
     if(!maxColors || !sigRGBbits || !loadPalette)
 	return FALSE;
