--- xorg-server-1.6.1.902/glx/glxdri2.c~	2009-06-29 23:10:40.000000000 +0200
+++ xorg-server-1.6.1.902/glx/glxdri2.c	2009-06-30 08:26:31.188770402 +0200
@@ -222,12 +222,15 @@
     if (texBuffer == NULL)
         return Success;
 
+#if __DRI_TEX_BUFFER_VERSION >= 2
     if (texBuffer->base.version >= 2 && texBuffer->setTexBuffer2 != NULL) {
 	(*texBuffer->setTexBuffer2)(context->driContext,
 				    glxPixmap->target,
 				    glxPixmap->format,
 				    drawable->driDrawable);
-    } else {
+    } else
+#endif
+    {
 	texBuffer->setTexBuffer(context->driContext,
 				glxPixmap->target,
 				drawable->driDrawable);
